### PR TITLE
Make key property mandatory in CREATE TABLE statement.

### PR DESCRIPTION
**Pull request by hjafarpour**
When we define a table using CREATE TABLE statement we should specify the key column using 'key' property in WITH clause.

### CODE REVIEW

**Comment by apurvam** [comment]
What was the key before this change? Was it deduced from the ordering of the fields in the DDL statement?

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java [hunk 1]
```diff
@@ -36,6 +37,11 @@ public CreateTableCommand(CreateTable createTable, Map<String, Object> overridde
 
     Map<String, Expression> properties = createTable.getProperties();
 
+    if (!properties.containsKey(DdlConfig.KEY_NAME_PROPERTY)) {
```
Comment (apurvam) [inline ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java]: Do we need to do additional validation? Like what if the key of the message has a different field name? What if  this key refers to a field in the value? Will that work?

File: ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java [hunk 2]
```diff
@@ -78,14 +79,31 @@ public void shouldCreateCommandForCreateStream() {
...
         new CreateTable(QualifiedName.of("foo"),
-            Collections.emptyList(), true, properties),
+            Collections.emptyList(), true, tableProperties),
         Collections.emptyMap());
 
     assertThat(result, instanceOf(CreateTableCommand.class));
   }
 
+  @Test
+  public void shouldFailCreateTableIfKeyIsNotProvided() {
```
Comment (apurvam) [inline ksql-engine/src/test/java/io/confluent/ksql/ddl/commands/CommandFactoriesTest.java]: Shouldn't we add tests which check the expected results when the explicitly specified key field doesn't exist is not the same as the key in the message.

**Comment by hjafarpour** [comment]
@apurvam before this we did not enforce setting the key and key would be null if it was not set. Now user has to set the key for Table. I added some extra checks to make sure that the specified column name in WITH clause for key or timestamp exists in the schema too.

File: ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java (hunk 1 above)
Comment (hjafarpour) [reply ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/CreateTableCommand.java]: I added additional check to make sure the specified key column name exists in the schema.

**Comment by blueedgenick** [comment]
This seems like a negative step from a UX perspective - more things to learn, which are non-intuitive, which add no observable improvement in behavior. 
At least in the version prior to this change the specification of 'key' is optional and joins seem to work without it (albeit with the performance penalty of mandatory re-keying). What's really strange about this strange from a user perspective, iiuc, is that we are essentially forcing a constraint on the user as to how they can structure their data to use it with ksql - i.e. the data of the message-key must be duplicated in the message-value and then declared in the value-schema and then referenced as this special 'key' thing, right ? or is my understanding out of date ?

**Comment by hjafarpour** [comment]
@dguy at the moment we require the changelog topic for the table to have the key that is specified by the user and it's user's responsibility to make sure that the data is provided correctly. As you mentioned repartitioning tables is not desired and that is the reason we support repartitioning (rekeying) only for streams and not for tables. 
@bluemonk3y we need to have the key set for Table since we will use this when we interpret the topic as a changelog and create ktable. Indeed not specifying the key for table will result in indeterministic behavior and side effects as @dguy said. So if the key is not set we cannot use the topic to create a ktable correctly.

**Comment by dguy** [comment]
if the key always has to be the record key then perhaps we don't need users to specify a key, but rather we need a "special" field that is used for joining on. The problem is if we add this change then people will think they can make the key any of the fields, which won't work.

**Comment by hjafarpour** [comment]
@dguy yes, I also agree that this may cause mistakes by users and they may think they can set the key to any column and expect correct results. But I think this is necessary for correctness. 
Also the reason to need the key to be one of the columns in the schema is that when users write join queries for instance they specify join criteria based on the defined schema that they have. For instance in the following query `userid` is already in the schema of both `pageview` and `users`:

`SELECT p.userid, pageid, region, gender FROM pageview p LEFT JOIN users u ON p.userid = u.userid;`

So inevitably we will have `userid` duplicated in both key and value for both `pageview` and `users` topics.

**Comment by apurvam** [comment]
For my understanding, what is today's behavior when we do a join against a `KTable` whose source topic (which is interpreted as the changelog topic) has a null key? And what if the key of the changelog topic is not the key we are joining against? Do we repartition the source topic to get the right key and then use the resulting KTable for the join? Or do we silently just fail to do joins?

If we do repartition, what value does this patch add exactly? If we don't repartition, then raising an error would be better than silently failing joins.

**Comment by hjafarpour** [comment]
@apurvam Currently in Stream-Table join if the stream is not keyed with the join criteria, we automatically repartition the stream to be able to perform the join. We require the table to be keyed with the join criteria since we cannot reparation a table. Repartitioning a table will result in possible incorrect results. So if we consider a topic as a changelog topic and read it as a Table (ktable) we should require the key to be specified and user cannot repartition the changelog topic to specify a desired column as the key for the table.
Finally, in the current implementation setting key column for table is optional which may result in incorrect behavior. The example you mentioned where the key is null will result in dropping all the messages in the table.

**Comment by hjafarpour** [comment]
retest this please

**Pull request merged**